### PR TITLE
Ignore geogram for MacOS

### PR DESCRIPTION
**Pull request by maurock**
This PR solves compilation errors with macOS M1/M2 (Issue #50 ). The errors are due to compatibility issues between the `geogram` library and the M1 architecture.

### CODE REVIEW

**Comment by fwilliams** [comment]
This looks great! I think we actually want to check if the target processor is ARM rather than if the system is Darwin. This will let us compile for general ARM architectures and close #41 as well as still support geogram functions on intel macs.

For the python checks, these can be replaced with [`platform.machine()`](https://docs.python.org/3/library/platform.html#platform.machine) checks (check if it matches `'arm'`). I'm not 100% sure how to do this in CMake but some googling will probably help!

**Comment by maurock** [comment]
You are right, I had a check on macOS rather than arm64 to be conservative, as I was not completely sure it worked on intel Mac. But thinking about it, why shouldn't it? I'll update it right away.

**Comment by fwilliams** [comment]
Great! Could you also add these changes so the CI generates arm builds?

https://github.com/fwilliams/point-cloud-utils/pull/49/files#diff-e1e84b3db32e874910839019d02caf72195920a29eed9c40be2e81cf4c8efeeaL29

**Comment by maurock** [comment]
I have added ARM64 for Mac, Linux, and Windows to the CI config. But there are a few issues with my latest changes:
- Check on the architecture in cmake is not working correctly, it is probably a syntax error.
- [This](https://github.com/fwilliams/point-cloud-utils/pull/55/commits/1199c9aaecfd07bc1fb165c8df13b6f5ab1bc845#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aL165) is incorrect because the target "_pcu_internal" was already created in the previous block

**Comment by maurock** [comment]
Re my previous comment:
1. Yes, it was just a silly mistake, I was checking for amd64 rather than arm64.
2. I have conditionally added `src/lloyd.cpp` as a binding source to the `npe` module when the arch is not arm64.

It works fine on the M1 and it should work on the other systems as well! :)

**Comment by fwilliams** [comment]
Thank you so much @maurock I'm running the CI now with your changes. If everything goes green, I think we can merge 🚀

File: CMakeLists.txt [hunk 1]
```diff
@@ -3,7 +3,9 @@ project(point-cloud-utils)
 set(CMAKE_CXX_STANDARD 14)
 set(CMAKE_BUILD_TYPE "Release")
 
-option(GEOGRAM_WITH_HLBFGS "Non-linear solver (Yang Liu's HLBFGS)"             ON)
+if(NOT ${CMAKE_SYSTEM_PROCESSOR} MATCHES "arm64")
```
Comment (fwilliams) [inline CMakeLists.txt:6]: below the check is for `"(arm64)|(ARM64)"`. Do we want the same check here? Maybe also `(arm32)|(ARM32)`?

**Comment by fwilliams** [comment]
@maurock looks like arm64 on Windows isn't building but it doesn't look like a problem with your code but the CI itself. Can you remove Windows arm64 builds from the CI?

**Comment by fwilliams** [comment]
Nevermind, I can edit the PR directly 🙂 

Let's see if this works!

**Comment by maurock** [comment]
We're so close! I've built it with my windows machine x86_64 and it seems like the error has to do with embree:
```
  CMake Error at cmake/DownloadProject.cmake:179 (message):
    Build step for embree failed: 1
  Call Stack (most recent call first):
    cmake/DownloadExternalDeps.cmake:13 (download_project)
    CMakeLists.txt:88 (download_dep)


  -- Configuring incomplete, errors occurred!
  See also "C:/Users/mauro/AppData/Local/Temp/pip-req-build-5m644mrs/build/temp.win-amd64-cpython-38/Release/CMakeFiles/CMakeOutput.log".
  See also "C:/Users/mauro/AppData/Local/Temp/pip-req-build-5m644mrs/build/temp.win-amd64-cpython-38/Release/CMakeFiles/CMakeError.log".
  <string>:34: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
  64 bit!
  Traceback (most recent call last):
    File "D:\Anaconda3\envs\interactive_ai\lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py", line 349, in <module>
      main()
    File "D:\Anaconda3\envs\interactive_ai\lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py", line 331, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "D:\Anaconda3\envs\interactive_ai\lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py", line 248, in build_wheel
      return _build_backend().build_wheel(wheel_directory, config_settings,
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\build_meta.py", line 413, in build_wheel
      return self._build_with_temp_dir(['bdist_wheel'], '.whl',
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\build_meta.py", line 398, in _build_with_temp_dir
      self.run_setup()
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\build_meta.py", line 335, in run_setup
      exec(code, locals())
    File "<string>", line 116, in <module>
    File "<string>", line 89, in main
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\__init__.py", line 87, in setup
      return distutils.core.setup(**attrs)
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\_distutils\core.py", line 185, in setup
      return run_commands(dist)
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\_distutils\core.py", line 201, in run_commands
      dist.run_commands()
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 969, in run_commands
      self.run_command(cmd)
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\dist.py", line 1208, in run_command
      super().run_command(command)
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 988, in run_command
      cmd_obj.run()
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\wheel\bdist_wheel.py", line 325, in run
      self.run_command("build")
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\_distutils\cmd.py", line 318, in run_command
      self.distribution.run_command(command)
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\dist.py", line 1208, in run_command
      super().run_command(command)
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 988, in run_command
      cmd_obj.run()
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\_distutils\command\build.py", line 132, in run
      self.run_command(cmd_name)
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\_distutils\cmd.py", line 318, in run_command
      self.distribution.run_command(command)
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\dist.py", line 1208, in run_command
      super().run_command(command)
    File "C:\Users\mauro\AppData\Local\Temp\pip-build-env-ild7u1av\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 988, in run_command
      cmd_obj.run()
    File "<string>", line 39, in run
    File "<string>", line 71, in build_extension
    File "D:\Anaconda3\envs\interactive_ai\lib\subprocess.py", line 364, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['cmake', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=C:\\Users\\mauro\\AppData\\Local\\Temp\\pip-req-build-5m644mrs\\build\\lib.win-amd64-cpython-38\\point_cloud_utils', '-DPYTHON_EXECUTABLE=D:\\Anaconda3\\envs\\interactive_ai\\python.exe', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_RELEASE=C:\\Users\\mauro\\AppData\\Local\\Temp\\pip-req-build-5m644mrs\\build\\lib.win-amd64-cpython-38\\point_cloud_utils', '-A', 'x64', 'C:\\Users\\mauro\\AppData\\Local\\Temp\\pip-req-build-5m644mrs']' returned non-zero exit status 1.
  ----------------------------------------
  ERROR: Failed building wheel for point-cloud-utils
```

**Comment by fwilliams** [comment]
hmm this is strange, it seems it's not able to clone in embree as a dependency (the Windows CI builds). I made some changes on the tip of your branch last night. Can you pull those and check?

It looks like we're almost there. The linux CI is not configured properly for arm builds, This is probably just me not knowing how to use CIBuildWheel effectively. It would be nice to generate arm builds on Linux too, so I'll look into this

**Comment by fwilliams** [comment]
This has turned into an adventure in updating dependencies. NumpyEigen (which PCU depends on) is depending on an old fork of pybind11, which I've had to update and backport my changes to. This hopefully fixes some of the compilation issues and gets us python3.11 support too!

**Comment by maurock** [comment]
Hi @fwilliams, thank you so much for doing this. Currently I am preparing for a workshop so I've been a bit slow in reacting. I see that it could not build the wheels for linux arch64 , I'll have a look tomorrow after work (UK time) and see if there's anything I can do to help.
Also, I am not sure why it could not clone in embree on windows. Is it because I added the x86_64 architecture on windows CI?

**Comment by Layer3** [comment]
I'm a bit out of the loop on this one, but if I can help with anything specific please just point me to it :)

**Comment by fwilliams** [comment]
Hey @Layer3! Thanks for popping in! We would like to integrate your simde changes into this branch (which updates a bunch of dependencies and does some small checks). I think if we do this, we should be able to build on M1 macs and Linux arm

**Comment by Layer3** [comment]
Okay, so the change was pretty straight forward.
Only caveats are that you'd have to deactivate aliases in simde so it completely overshadows the normal sse instructions and then there was one (igpl or sth?) that had a platform switch that needs handling as well and I think that's about it.
That being said, this can lead to problems if you try to build on a platform supporting sse *while* using simde, which I wouldn't recommend.
To really commit to simde on every platform would require a bit more work.

**Comment by fwilliams** [comment]
I don’t think we need simde on every platform. Only enabling it on arm builds is fine. I patched libigl to not use SSRI instructions for now so I think we only need the changes for geogram and the platform checks to make sure we use SSE on Intel builds and simde on arm.

**Comment by fwilliams** [comment]
Looks like the CI Builds on MacOS arm64 and Linux arm64 (though for some reason Linux arm builds take an eternity).

With @Layer3 's changes we can bring back geogram on Mac. I'm going to make some edits to this PR and merge it. For now Linux ARM64 users will have to build from source to keep the CI running at a reasonable speed.

**Pull request merged**